### DISCUSSION TITLE
Importance of sorting when using multiple transaction files

### DISCUSSION THREAD

**Discussion by jacobyavis** [Open-Cap-Table-Coalition/Open-Cap-Format-OCF#503] [General]
**Background:**
* In early days of OCF, the file format did not use a ZIP structure. Rather the entire format was housed in a single .json object
* The introduction of the ZIP architecture and support for multiple files was aimed to address a possible issue with the OCF .json file being too large (i.e. too large to transmit easily, too large to process in the memory of some applications)
** Now I admit, the problem was largely theoretical. In practice would we ever have a cap table so large that the data can't actually fit within the memory constraints of our modern hardware?
** But regardless, theoretical or not, it was deemed important enough to make the change and introduce the ZIP structure.

**Today:**
* I've heard that the validation tool (being worked on by the group led by @arthur-clara) loads ALL transactions files in memory and sorts them before doing any processing
* If one group of people are making this assumption, then others are sure to do the same... right?
* Which means that we have a classic case of missing the point. What was the point of splitting the arch into multiple transaction files if certain tools/implementations we're building must load them ALL before doing any processing?

**Order Important?**
The [OCF Architecture](https://open-cap-table-coalition.github.io/Open-Cap-Format-OCF/explainers/Architecture/) is described as a stack architecture. It's inherent that order matters when using a Stack data structure. But are we generally in agreement that a given OCF implementation needn't worry about ordering the transactions within the transactions file(s)? It's a logical impediment to make assumptions.

Perhaps within a single transactions file, the order needn't matter. But in the OCF documentation (on the [OCFManifestFile](https://open-cap-table-coalition.github.io/Open-Cap-Format-OCF/schema_markdown/schema/files/OCFManifestFile/) docs) we state a sentence like this for each of the arrays of file references:

 > List of files containing lists of issuer transactions, **indexed from the file containing the first such object created to the file containing the last**
 
 Meaning OCF expected to be able to process the files one-by-one, and for that to be possible, we need sorting to be in logical order.

Opening this up for discussion as there seems to be some confusion on the topic. If order truly doesn't matter (and the size problem is indeed theoretical only) then we could return to a simplified OCF JSON format instead of the multi-file architecture.

**Comment by arthur-clara** [comment]
@jacobyavis 

This is a good point and valid that order matters in a Stack data structure. 

If we can ensure that ordering is done either by file or within files, it will be better than having to do a sort on all transactions. I agree that it may be largely academic, but it is still best practice.

@pjohnmeyer @JSv4 Can we discuss at the next TWG and decide a proposal. If we are enforcing ordering of some sort, then we would need to validate that ordering is being adhered to. Either through the OCF schema validator or in the validation tool itself. 

And it may be that everyone else is ordering and I am the only one assuming that a transaction file may or may not be ordered in which case we should validate ordering.

**Reply by pjohnmeyer** [reply]
The `ocx` software operated under the assumption that data would not be ordered, but we didn't get very far into what that meant. It was designed to cache out of order info until it wasn't needed anymore. It would, however, operate more efficiently (primarily in terms of memory) if the data was presented in an ordered fashion. The assumption was that another tool/library would be able to sort TXNs and stream them in.

Ultimately I don't have a dog in this fight, but I personally would leave ordering unspecified in the standard.
